rust/tools/authorization-logic/README.md

```diff
@@ -0,0 +1,209 @@
+using inference rules and facts written in this language. Generally, 
+authorization logic is paired with a reference monitor that interprets the 
+policy language and permits the guarded actions only when presented with a 
+proof of the proposition used to grant access.
+
+### Principals and decentralization
+The key feature of authorization logic is that facts are decentralized - each 
+fact or rule is attributed to a principal (or actor). For example, HR says 
+isManager(Alice) means that HR believes or claims that isManager(Alice) is 
+true. Decentralizing facts in this way ensures that a resource owner retains 
+control over access to resources in an environment with mutually distrusting 
+principals. As an example, to use the rule
+```
+"FileServer" says canAccess(file, requestor) :- isManager(requestor),                
+                                              isValidFile(file).
+```
+requires proving that FileServer says isManager(requestor) and FileServer says 
+isValidFile(file). 
+
+### Delegation
+Because authorization logic models beliefs held by various parties rather than 
+a single global truth believed by all, we need a way to export facts from one 
+principal to another. In authorization logic, this is done by delegation. For 
+example:
+
+```
+FileServer says HR canSay isManager(x).
+```
+means that HR canSay isManager(x) implies FileServer says isManager(x). The 
+exact language-level construct for delegation has varied among authorization 
+logics, but the key point is that this gives one principal (FileServer) a way 
+of specifying when it believes claims made by another (HR). Delegations can 
+also be done conditionally by placing them on the left of a proof rule.
+
+
+### Authenticating policies
+When FileServer imports a policy statement from HR, we want to know that a 
+claim genuinely originated from HR. To authenticate policy statements, each 
+principal is also associated with an asymmetric key-pair. When a policy 
+statement is exported by a principal, the statement is digitally signed, the 
+signature is sent to the recipient along with a serialization of the policy 
+claim, and the recipient uses the public key to check the signature.
+
+## Instructions
+
+### Running a program
+To compile a program, run:
+```
+cargo run FILENAME INPUT_DIR OUTPUT_DIR
+```
+which will result in a souffle file (with the extension .dl) in OUTPUT\_DIR. If 
+Souffle is installed, it will be run on the generated code.
+
+### Language Guide
+
+#### Statements, variables, and constants
+Programs are sequences of assertions that are attributed to principals which we 
+call statements. Statements can either be _facts_ of the form
+```
+<principal_name> says <predicate_name>(arg_1, ..., arg_n).
+```
+meaning that the principal believes predicate (applied to its arguments) to be true, or
+_rules_ of the form
+```
+<principal_name> says <predicate_name>(arg_1, ..., arg_n) :- 
```

Comment:
It might be useful to include information on what characters can be included in principle name, variable names etc. Also, can quoted values contain more characters than names? Does some characters require escaping?